### PR TITLE
feature: Sponsors admin & API functionality

### PR DESCRIPTION
**Pull request by adamoxley**
# Changes

- Adds in admin sponsor (create)
- Adds in admin API sponsor functionality (create, edit)
- Adds in new ImageService class for uploading to S3
- Adds in new `S3_BUCKET_NAME` environment (note: this will need added to CI for prod if we choose to use it, IMO, it's good for local dev to have it in)
- Updates website to use new sponsor API

# Notes

- SVG seem to not work for some reason, I think due to how the S3 uploads are done, the SVG files appear correct (and download correct) directly from the S3 bucket, but don't show in Safari or Chrome. I suggest we use PNG's going forward, as these would render on iOS just fine also.
- The way sponsors are done right now is passing in each individual sponsor level into `HomeContext`, I've done this due to Leaf templates not supporting dictionaries and not sure what other way to approach it. Suggestions welcome.

# Screenshots

## Create Sponsor

Screenshot 2022-07-21 at 22 00 21

## Create Sponsor (sponsor level section)

Screenshot 2022-07-21 at 00 13 36

## Admin Sponsor List

Screenshot 2022-07-21 at 22 00 06

## Edit Sponsor

Screenshot 2022-07-21 at 21 59 23

## Website Sponsor List

Screenshot 2022-07-21 at 00 12 46

### CODE REVIEW

File: Resources/Views/Home/_speakers.leaf [hunk 1]
```diff
@@ -24,7 +24,7 @@
         #for(speaker in speakers):
             <div class="speaker-grid-element">
                 <div class="speaker-grid-element-image-container">
-                    <img class="speaker-grid-element-image" src="https://swiftleeds-speakers.s3.eu-west-2.amazonaws.com/#(speaker.profileImage)"/>
+                    <img class="speaker-grid-element-image" src="https://swift-leeds.s3.eu-west-2.amazonaws.com/#(speaker.profileImage)"/>
```
Comment (adamoxley) [inline Resources/Views/Home/_speakers.leaf:27]: Whoops. This is my local bucket. Will swap it back before it gets merged in.

**Review by SwiftyAlex — CHANGES_REQUESTED** [review]
Just the one thought, looks great!

File: Sources/App/Features/Sponsors/Models/Sponsor.swift [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    }
+    
+    @ID(key: .id)
+    var id: UUID?
+    
+    @Field(key: "name")
+    var name: String
+    
+    @Field(key: "image")
+    var image: String
```
[the remark this answers is not in the document]
Comment (adamoxley) [reply Sources/App/Features/Sponsors/Models/Sponsor.swift:26]: So I couldn't get web to display uploaded SVG's. I think its the way we use `S3.PutObjectRequest` but I'm honestly not sure. They uploaded fine into the S3 bucket and I could download from the S3 bucket and view them locally, but they never displayed on the web using the same `img` code as was in using local SVG's. 

I'd say if we use PNG's for now it's probably _fine_ for both platforms?

**Comment by adamoxley** [comment]
@adamrushy @SwiftyAlex @pdamonkey 

Ready for review.

I haven't included the mobile API into this PR as I think that needs some discussion amongst the team as to how we're going to structure the mobile APIs away from the admin APIs which route under `api/v1` (I think if I'm remembered correctly). But this at least lays all the ground work for the sponsor API.

Updated the main PR description with the latest screenshots of the website & admin panel. Included in the edit sponsor web page is a preview of the existing sponsors image, as it isn't possible to pre-populate a file input field, if no file is uploaded on edit sponsor, the old image is kept.

Before merging this in we will need to ensure two things;
- A new env key gets added to hosting for S3_BUCKET_NAME
- We have all the sponsor images / other info ready to re-upload. Images should be in PNG format (a few sponsors have supplied them as SVG).

File: Resources/Views/Home/_speakers.leaf (hunk 1 above)
Comment (adamoxley) [reply Resources/Views/Home/_speakers.leaf]: I've changed back but GitHub isn't seeing this as an outdated comment for some reason. Will resolve conversation instead.

File: Sources/App/Features/Sponsors/Models/Sponsor.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (SwiftyAlex) [reply Sources/App/Features/Sponsors/Models/Sponsor.swift:26]: very weird! ok, lets rock with these for now, we'll just have to make sure those pngs are _crispy_ 👍🏻

**Pull request merged**